### ISSUE TITLE
Wrong route priority with asterisk match

### ISSUE THREAD

**Issue by drogus** [emberjs/ember.js#10025]
I've came across a potential bug in the router. Given the following `map` definition:

```javascript
App.Router.map(function() {
  this.route('search', { path: '/search/*phrase' });
  this.route('other', { path: '/:foo/:bar' });
});
```

it's impossible to reach `search` route when entering through `/search/a-search-phrase` URL. Here is a jsbin demonstrating this: http://emberjs.jsbin.com/waxovu/1/edit?html,js,console,output and here is what happens when you try to enter search route: http://emberjs.jsbin.com/waxovu/1#/search/something. If you enter this URL, you should see `Other template: {"foo":"search","bar":"something"}`. Links correctly transition to both routes.

I'm happy to dig into this further and fix it, but I wanted to get confirmation that this is a proper code and not my misunderstanding of how things work with asterisk match.

**Comment by drogus** [comment]
It seems that the problem is caused by the [`sortSolutions` function](https://github.com/tildeio/route-recognizer/blob/master/lib/route-recognizer.js#L241) in route-recognizer. A comment on top of the function mentions what's a general strategy, but there is no mention what's the purpose of this in the first place. My first intuition when using router was that routes priority is based on the order of adding them, so such a router:

```javascript
Router.map(function() {
  this.route('search', { path: '/search/*phrase' });
  this.route('repo', { path: '/:owner/:name' });
});
```

would match the `search` route first and then `repo` route if search was not matched.

I think that such a behaviour is easier to document and easier to predict for a developer. I suppose changing that in 1.x release is a no-go for backwards compatibility, but maybe it could be addressed in 2.0?

**Comment by wagenet** [comment]
@drogus can you re-file here then? https://github.com/tildeio/route-recognizer
